app/routes.py

```diff
@@ -1,2 +1,238 @@
+    message_to_slack("#task-notifications", f"Someone just completed the task {task.title}")
+
+    return {
+        "task": task.to_json()
+    }, 200
+
+#creating custom endpoints with mark_incomplete: True or False
+@tasks_bp.route("/<task_id>/mark_incomplete", methods=["PATCH"]) #update
+def mark_incomplete(task_id):
+    task = Task.query.get(task_id)
+    if not task:
+        return ("", 404)
+
+    if task.completed_at != None:
+        task.completed_at = None
+    
+    db.session.commit()
+
+    return {
+        "task": task.to_json()
+        }, 200
+
+#checks wave_4 
+#created a function to send API message to slack 
+def message_to_slack(channel, message):
+
+    token = os.environ.get("SLACK_API_TOKEN") #SLACK_API_TOKEN from .env file 
+
+    return requests.post('https://slack.com/api/chat.postMessage', headers = {
+        "Authorization": f"Bearer #{token}"
+    },
+    data = { 
+        "channel": channel,
+        "text": message
+    })
+
+
+#checks wave_5 tests
+#created second models for goals
+@goals_bp.route("", methods=["POST"])
+def create_goal():
+    request_body = request.get_json()
+
+    if "title" not in request_body:
+        return (
+            {"details": "Invalid data"}
+            ), 400
+    else:
+        add_goal = Goal(title = request_body["title"])
+
+        db.session.add(add_goal)
+        db.session.commit()
+
+        return make_response({
+                "goal": add_goal.to_json_goal()
+            }, 201)
+
+@goals_bp.route("", methods=["GET"])
+def get_goals():
+    goals = Goal.query.all()
+
+    goals_response = []
+    for goal in goals: 
+        goals_response.append(goal.to_json_goal())
+
+    return jsonify(goals_response)
+
+@goals_bp.route("/<goal_id>", methods=["GET", "PUT", "DELETE"])
+def handle_goals(goal_id):
+
+    goal = Goal.query.get(goal_id)
+    if goal is None:
+        return make_response("", 404)
+    
+    if request.method == "GET":
+        return make_response(
+            {"goal": goal.to_json_goal()
+        }, 200)
+
+    elif request.method == "PUT":
+        form_data = request.get_json()
+        goal.title = form_data["title"]
+
+        db.session.commit()
+
+        return make_response(
+            {"goal": goal.to_json_goal()}, 200)
+
+    elif request.method == "DELETE":
+
+        db.session.delete(goal)
+        db.session.commit()
+
+        return make_response({
+            "details": f"Goal {goal.goal_id} \"{goal.title}\" successfully deleted"
+            })
+
+
+#Wave 6
+#Establishing a One-to-Many Relationship between goals and tasks
+@goals_bp.route("/<goal_id>/tasks", methods=["POST"])
+def create_goals_tasks(goal_id):
+    request_body = request.get_json()
+
+    number = int(goal_id)
+    for task_id in request_body["task_ids"]:
+        task = Task.query.get(task_id)
+
+        task.goal_id = number
+
+    db.session.commit()
+    return {
+        "id": number, 
+        "task_ids": request_body["task_ids"]
+    }
```

Comment:
This change would have made the above-mentioned bug more obvious. This would also allow the response to include ALL tasks associated with the goal, not just the new ones.
This isn't super intuitive so I'd definitely be down to talk through this more in our next 1:1

```suggestion
    full_task_ids = []
    for task in goal.tasks:
        full_task_ids.append(task.task_id)
        
    return {
        "id": number, 
        "task_ids": full_task_ids
    }
```